### ISSUE TITLE
Lint/UselessAssignment vs postfix if

### ISSUE THREAD

**Issue by akimd** [rubocop/rubocop#12269]
Hi!

## Expected behavior

```ruby
puts "there were #{nerrs}" if (nerrs = get_nerrs) != 0
```

The assignment into `nerrs` is clearly used.

## Actual behavior

```
$ bundle exec rubocop nerrs.rb
Inspecting 1 file
E

Offenses:

nerrs.rb:1:1: C: [Correctable] Style/FrozenStringLiteralComment: Missing frozen string literal comment.
puts "there were #{nerrs}" if (nerrs = get_nerrs) != 0
^
nerrs.rb:1:32: E: [Correctable] Lint/UselessAssignment: Useless assignment to variable - nerrs.
puts "there were #{nerrs}" if (nerrs = get_nerrs) != 0
                               ^^^^^

1 file inspected, 2 offenses detected, 2 offenses autocorrectable
```

## Steps to reproduce the problem

Test case above.

## RuboCop version

```
$ bundle exec rubocop -V
1.56.3 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [arm64-darwin22]
```

**Comment by koic** [comment]
This doesn't seem to be a bug. When using a postfix `if`, there's an error when referencing `nerrs`.

```console
$ ruby -e 'puts "there were #{nerrs}" if (nerrs = true)'
-e:1: warning: found `= literal' in conditional, should be ==
(none):1: warning: found `= literal' in conditional, should be ==
-e:1:in `<main>': undefined local variable or method `nerrs' for main (NameError)

puts "there were #{nerrs}" if (nerrs = true)
                   ^^^^^
```

To achieve the expected behavior, it might be necessary to write a regular `if` statement.

```ruby
if (nerrs = true)
  puts "there were #{nerrs}"
end
```

**Comment by akimd** [comment]
Hi!

> When using a postfix if, there's an error when referencing nerrs.

Oh great!  I so strongly dislike this coding style...

Thanks a lot for the very fast feedback.

**Comment by akimd** [comment]
Well, that story was too beautiful.  It's only when the variable is **introduced** in the postfix-if that Ruby does not run the assignment.  If the variable is defined before, it does run the assignment before.

And in that case, RuboCop is still wrong about a useless assignment.  But this time is tags the definition of the variable.

```
$ cat nerrs.rb
def get_nerrs
  23
end
nerrs = 0
raise "there were #{nerrs}" if (nerrs = get_nerrs) != 0
```
```
$ ruby nerrs.rb
nerrs.rb:5:in `<main>': there were 23 (RuntimeError)
```
```
$ bundle exec rubocop nerrs.rb
Inspecting 1 file
E

Offenses:

nerrs.rb:1:1: C: [Correctable] Style/FrozenStringLiteralComment: Missing frozen string literal comment.
def get_nerrs
^
nerrs.rb:1:5: C: Naming/AccessorMethodName: Do not prefix reader method names with get_.
def get_nerrs
    ^^^^^^^^^
nerrs.rb:4:1: E: [Correctable] Lint/UselessAssignment: Useless assignment to variable - nerrs.
nerrs = 0
^^^^^

1 file inspected, 3 offenses detected, 2 offenses autocorrectable
```

**Comment by koic** [comment]
I see! That case is a false positive, I think. Thank you for looking into it.

**Comment by akimd** [comment]
This second example features another failure: `Lint/SelfAssignment` does not see that `nerrs = nerrs` is not a self assignment (it actually stands for `nerrs = self.nerrs`).  And again it thinks the assignment is useless, but it is not.

```
$ cat nerrs.rb
class Foo
  def nerrs
    puts "Foo#nerrs"
    23
  end

  def run
    puts "there were #{nerrs} errors" if (nerrs = nerrs) != 0
  end
end

Foo.new.run
```
```
$ ruby nerrs.rb
Foo#nerrs
there were 23 errors
```
```
$ bundle exec rubocop nerrs.rb
Inspecting 1 file
E

Offenses:

nerrs.rb:1:1: C: [Correctable] Style/FrozenStringLiteralComment: Missing frozen string literal comment.
class Foo
^
nerrs.rb:3:10: C: [Correctable] Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    puts "Foo#nerrs"
         ^^^^^^^^^^^
nerrs.rb:8:43: E: Lint/SelfAssignment: Self-assignment detected.
    puts "there were #{nerrs} errors" if (nerrs = nerrs) != 0
                                          ^^^^^^^^^^^^^
nerrs.rb:8:43: E: [Correctable] Lint/UselessAssignment: Useless assignment to variable - nerrs.
    puts "there were #{nerrs} errors" if (nerrs = nerrs) != 0
                                          ^^^^^

1 file inspected, 4 offenses detected, 3 offenses autocorrectable
```
